src/builtin.rs

```diff
@@ -18,15 +21,54 @@ fn get_builtin(cmd : &str) -> Option<&Builtin> {
     })
 }
 
+fn normalize_logical_path(path: &PathBuf) -> PathBuf {
+    let mut normalized_path = PathBuf::new();
+    for c in path.components() {
+        match c {
+            Component::ParentDir => { normalized_path.pop(); },
+            Component::CurDir => continue,
+            _ => normalized_path.push(c.as_os_str()),
+        };
+    }
+    normalized_path
+}
+
 pub fn cd(args : &[&str]) -> io::Result<i32> {
-    let path = Path::new(args[0]);
-    env::set_current_dir(path).and(Ok(0))
+    fn change_to(p: &PathBuf) -> io::Result<()> {
+        let new_pwd_buf = normalize_logical_path(p);
+        env::set_var("PWD", &new_pwd_buf);
+        env::set_current_dir(new_pwd_buf)
```

Comment:
You're passing `new_pwd_buf` to `set_current_dir`, won't it not come back from that?